CorsixTH/Lua/room.lua

```diff
@@ -573,16 +573,18 @@ function Room:onHumanoidLeave(humanoid)
     end
     -- Make patients leave the room if there are no longer enough staff
     if not self:testStaffCriteria(self:getRequiredStaffCriteria()) then
-      local patient_needs_to_reenter = false
+      local call_for_new_staff = false
       for room_humanoid in pairs(self.humanoids) do
-        if class.is(room_humanoid, Patient) and self:shouldHavePatientReenter(room_humanoid) then
-          self:makeHumanoidLeave(room_humanoid)
-          room_humanoid:queueAction(self:createEnterAction(room_humanoid))
-          patient_needs_to_reenter = true
+        if class.is(room_humanoid, Patient) and not room_humanoid:hasLeavingAction() then
+          call_for_new_staff = true
+          if self.room_info.id ~= "ward" then
+            self:makeHumanoidLeave(room_humanoid)
+            room_humanoid:queueAction(self:createEnterAction(room_humanoid))
+          end
         end
       end
       -- Call for staff if needed
-      if self.is_active and (patient_needs_to_reenter or self.door.queue:patientSize() > 0) then
+      if self.is_active and (call_for_new_staff or self.door.queue:patientSize() > 0) then
```

Comment:
You can replace `false` at line 576 by `(self.door.queue:patientSize() > 0)`  so you don't need it here.